File: src/main/java/macro303/neptunes/Team.java
package macro303.neptunes;

import macro303.neptunes.player.Player;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

/**
 * Created by Macro303 on 2018-05-10.
 */
public class Team implements Comparable<Team> {
	@NotNull
	private final ArrayList<Player> members = new ArrayList<>();

	public boolean isLost() {
		return members.stream().allMatch(Player::isAI);
	}

	@NotNull
	public String getName() {
		return members.size() > 0 ? members.get(0).getTeam() : "Unknown";
	}

	public int getTotalShips() {
		int sum = members.stream().filter(member -> !member.isAI() && !member.isConceded()).mapToInt(Player::getTotalShips).sum();
		return sum;
	}

	public int getTotalStars() {
		return members.stream().filter(member -> !member.isAI() && !member.isConceded()).mapToInt(Player::getTotalStars).sum();
	}

	public int getTotalEconomy() {
		return members.stream().filter(member -> !member.isAI() && !member.isConceded()).mapToInt(Player::getTotalEconomy).sum();
	}

	public int getEconomyTurn() {
		return members.stream().filter(member -> !member.isAI() && !member.isConceded()).mapToInt(Player::getEconomyTurn).sum();
	}

	public int getTotalIndustry() {
		return members.stream().filter(member -> !member.isAI() && !member.isConceded()).mapToInt(Player::getTotalIndustry).sum();
	}

	public int getIndustryTurn() {
		return members.stream().filter(member -> !member.isAI() && !member.isConceded()).mapToInt(Player::getIndustryTurn).sum();
	}

	public int getTotalScience() {
		return members.stream().filter(member -> !member.isAI() && !member.isConceded()).mapToInt(Player::getTotalScience).sum();
	}

	public void addMember(Player player) {
		members.add(player);
	}

	@Override
	public int compareTo(@NotNull Team other) {
		if (getTotalStars() - other.getTotalStars() != 0)
			return getTotalStars() - other.getTotalStars();
		if (getTotalShips() - other.getTotalShips() != 0)
			return getTotalShips() - other.getTotalShips();
		if ((getTotalEconomy() + getTotalIndustry() + getTotalScience()) - (other.getTotalEconomy() + other.getTotalIndustry() + other.getTotalScience()) != 0)
			return (getTotalEconomy() + getTotalIndustry() + getTotalScience()) - (other.getTotalEconomy() + other.getTotalIndustry() + other.getTotalScience());
		return getName().compareToIgnoreCase(other.getName());
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Team)) return false;

		Team team = (Team) o;

		return members.equals(team.members);
	}

	@Override
	public int hashCode() {
		return members.hashCode();
	}

	@Override
	public String toString() {
		return "Team{" +
				"members=" + members +
				'}';
	}
}

File: src/main/kotlin/macro303/neptunes/team/Team.kt
package macro303.neptunes.team

import macro303.neptunes.player.Player
import java.util.*

/**
 * Created by Macro303 on 2018-05-10.
 */
class Team : Comparable<Team> {
	private val members = ArrayList<Player>()

	val isLost: Boolean
		get() {
			return members.stream().allMatch({ it.ai || it.conceded })
		}

	val name: String
		get() = if (members.size > 0) members[0].team else "Unknown"

	val totalShips: Int
		get() = members.stream().filter { member -> !member.ai || !member.conceded }.mapToInt({ it.totalShips }).sum()

	val totalStars: Int
		get() = members.stream().filter { member -> !member.ai || !member.conceded }.mapToInt({ it.totalStars }).sum()

	val totalEconomy: Int
		get() = members.stream().filter { member -> !member.ai || !member.conceded }.mapToInt({ it.totalEconomy }).sum()

	val economyTurn: Int
		get() = members.stream().filter { member -> !member.ai || !member.conceded }.mapToInt({ it.economyTurn }).sum()

	val totalIndustry: Int
		get() = members.stream().filter { member -> !member.ai || !member.conceded }.mapToInt({ it.totalIndustry }).sum()

	val industryTurn: Int
		get() = members.stream().filter { member -> !member.ai || !member.conceded }.mapToInt({ it.industryTurn }).sum()

	val totalScience: Int
		get() = members.stream().filter { member -> !member.ai || !member.conceded }.mapToInt({ it.totalScience }).sum()

	fun addMember(player: Player) {
		members.add(player)
	}

	override fun compareTo(other: Team): Int {
		if (totalStars.compareTo(other.totalStars) != 0)
			return totalStars.compareTo(other.totalStars)
		if (totalShips.compareTo(other.totalShips) != 0)
			return totalShips.compareTo(other.totalShips)
		if ((totalEconomy + totalIndustry + totalScience).compareTo(other.totalEconomy + totalIndustry + totalScience) != 0)
			return (totalEconomy + totalIndustry + totalScience).compareTo(other.totalEconomy + totalIndustry + totalScience)
		return name.compareTo(other.name, ignoreCase = true)
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is Team) return false

		if (members != other.members) return false

		return true
	}

	override fun hashCode(): Int {
		return members.hashCode()
	}

	override fun toString(): String {
		return "Team(members=$members)"
	}
}